loopy/schedule/tools.py

```diff
@@ -138,3 +140,215 @@ def get_return_from_kernel_mapping(kernel):
+    from loopy.kernel.data import (filter_iname_tags_by_type,
+                                   HardwareConcurrentTag)
+
+    gsize, lsize = knl.get_grid_size_upper_bounds(callables_table,
+                                                  return_dict=True)
+
+    # {{{ Step 1: Preprocess the maps
+
+    # Step 1.1: Project out inames which are also map's dims, but does not form the
+    #           insn's within_inames
+    # Step 1.2: Project out sequential inames in the access maps
+    # Step 1.3: Rename the dims with their iname tags i.e. (g.i or l.i)
+    # Step 1.4: Name the ith output dims as _lp_dim{i}
+
+    updated_maps = []
+
+    for (map_, insn) in [
+            (map_a, insn_a),
+            (map_b, insn_b)]:
+        for iname in ((set(map_.get_var_dict().keys()) & knl.all_inames())
+                      - insn.within_inames):
+            # project out all inames in the map's domain that aren't part of the
+            # insn's domain
+            dt, pos = map_.get_var_dict()[iname]
+            map_ = map_.project_out(dt, pos, 1)
+
+        for iname in insn.within_inames:
+            if not filter_iname_tags_by_type(knl.inames[iname].tags,
+                                             HardwareConcurrentTag):
+                dt, pos = map_.get_var_dict()[iname]
+                map_ = map_.project_out(dt, pos, 1)
+
+        for name, (dt, pos) in map_.get_var_dict().items():
+            if dt == isl.dim_type.in_:
+                tag, = filter_iname_tags_by_type(knl.inames[name].tags,
+                                                 HardwareConcurrentTag)
+                map_ = map_.set_dim_name(dt, pos, str(tag))
+
+        for i_l in lsize:
+            if f"l.{i_l}" not in map_.get_var_dict():
+                ndim = map_.dim(isl.dim_type.in_)
+                map_ = map_.add_dims(isl.dim_type.in_, 1)
+                map_ = map_.set_dim_name(isl.dim_type.in_, ndim, f"l.{i_l}")
+
+        for i_g in gsize:
+            if f"g.{i_g}" not in map_.get_var_dict():
+                ndim = map_.dim(isl.dim_type.in_)
+                map_ = map_.add_dims(isl.dim_type.in_, 1)
+                map_ = map_.set_dim_name(isl.dim_type.in_, ndim, f"g.{i_g}")
+
+        for pos in range(map_.dim(isl.dim_type.out)):
+            map_ = map_.set_dim_name(isl.dim_type.out, pos, f"_lp_dim{pos}")
+
+        updated_maps.append(map_)
+
+    map_a, map_b = updated_maps
+
+    # }}}
+
+    # {{{ Step 2: rename all lid's, gid's in map_a to lid.A, gid.A
+
+    for name, (dt, pos) in map_a.get_var_dict().items():
+        if dt == isl.dim_type.in_:
+            map_a = map_a.set_dim_name(dt, pos, name+".A")
+
+    # }}}
+
+    # {{{ Step 3: rename all lid's, gid's in map_b to lid.B, gid.B
+
+    for name, (dt, pos) in map_b.get_var_dict().items():
+        if dt == isl.dim_type.in_:
+            map_b = map_b.set_dim_name(dt, pos, name+".B")
+
+    # }}}
+
+    # {{{ Step 4: make map_a, map_b ISL sets
```

Comment:
Why is it necessary to convert these to sets? If it is for `isl_set_from_expr`, I suspect that can be modified trivially to return a `Map`.

Reply:
> Why is it necessary to convert these to sets?

I wasn't able to cast the final condition as an intersect operation on the maps.

> If it is for `isl_set_from_expr` 

That's not much important, writing an inequality constraint is a fairly small piece of code.

Reply:
> I wasn't able to cast the final condition as an intersect operation on the maps.

Shouldn't that just be the same condition? (Not a showstopper for merging, just curious.)